Fix opening VideoDetailFragment and much more

Stypox opened the pull request:
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- Unify all ways of opening the VideoDetailFragment into two methods: `openVideoDetailFragment()` which sends the new data directly to the fragment manager, `openVideoDetail()` which sends an `Intent` received by `MainActivity` which in turn calls `openVideoDetailFragment()`.
- The arguments taken by `openVideoDetailFragment()` have changed: `autoPlay` is not there anymore, and `switchingPlayers` was added to be able to always determine correctly if the function was called as a way to switch players. `autoPlay` is now calculated based on the user's preference, on `switchingPlayers` and on whether there is already a background/popup player running. For more info see [these lines](https://github.com/TeamNewPipe/NewPipe/pull/4562/files#diff-7c684dc26bc21808937b0cd7c85de9810279b15b2ba4fdd2e31a0ad2c9e94962R347-R361). This makes it far more clear when the mini player will start playing (I think some situations were handled correcly only by accident before)
- Disable player stream preloading only if the current stream is going to be replaced for sure (see [this](https://github.com/TeamNewPipe/NewPipe/pull/4562/files#diff-b2e369e39e24aea82f68bcb17acd1135173d776a871986ac8ef85ad73e767333R871-R875)). `equals()` was implemented for `PlayQueueItem`s, so that (only) the url is compared when checking them.
- Add `@Nullable` and `@NotNull` to some variables in `VideoDetailFragment`, allowing me to enforce or remove some null checks. Rename `name` to `title` for consistency.
- Remove `static` qualifier from `currentInfo` and `playQueue` in `VideoDetailFragment`: they were never accessed as static members, and having them static caused random problems while replacing the `VideoDetailFragment` with a new instance of itself.
- [Do not force reopen](https://github.com/TeamNewPipe/NewPipe/pull/4562/files#diff-7c684dc26bc21808937b0cd7c85de9810279b15b2ba4fdd2e31a0ad2c9e94962R510-R511) `MainActivity` when sharing something to NewPipe and pressing on "Show info". This was introduced in 67d2b9131e7871505454d12dd1d245528b251228 by @yausername while handling urls in comments, but at that point there was no unified player, so probably without that it didn't work. But now everything works good even by keeping the current `MainActivity` instance, and this yields many improvements: opening urls in descriptions in NewPipe now does not erase backstack, nor does clicking `Open details` on queue items, and probably something else. :-D 
- Replace "Video player" with "Show info" when sharing a video to NewPipe but the background or popup players are running, and thus `autoPlay` would be `false`.
- Handle "Video player" in the exact same way as "Show info" for streams, i.e. do not start a service just to fetch the info, since `VideoDetailFragment` can fetch data itself without issues.
- When sharing a video to NewPipe and a play button is pressed, replace the current queue instead of enqueueing. Enqueueing can still be achieved opening info and then enqueueing, but enqueueing directly was bad UX since the user would not see what happened.
- When opening NewPipe from scratch but a player is already running, show the bottom mini player #4712
- Make sure play/pause state is preserved when switching players via the specific buttons. Also replace `START_PAUSED` with `PLAY_WHEN_READY` for better consistency (i.e. not inverting twice).
- Move menu listener to switch players from `BackgroundPlayerActivity` to `ServicePlayerActivity`, thus removing a useless abstract function.
- Remove [some unused code](https://github.com/TeamNewPipe/NewPipe/pull/4562/files#diff-01c2c8e04aff70b98bd2a94267e903d313f32a2e1df712a023e56ace95c9bf62R556-R557) in `RouterActivity`
- Refactor `getPlayerIntent()` functions in `NavigationHelper` and removed unused navigation functions.
- Remove testing-only code in [PlayQueue](https://github.com/TeamNewPipe/NewPipe/pull/4562/files#diff-00e000f7dc0e535b7d97ab5d6056c88e7191a7480725fb20ccd4c49e36b4aabc) `reportingReactor` was used to capture and log broadcast events in debug builds. This can be undone if unwanted.
- Remove unused preference `autoplay_through_intent_key` (left there by accident after 2907, probably)
- Some refactoring here and there: fix warnings in `openDownloadDialog`, in `openDownloads`, in `PlayerHelper` and more

#### Fixes the following issue(s)
Fixes #4541 
Probably fixes #4432
Fixes #4403
Fixes (maybe) #4687
Fixes #4712

#### APK testing 
Since this PR changes the logic of many things, the more people can test, the better.
@opusforlife2 @Jean757 @shadow00 @nbmrjuhneibkr could you test if the new share behaviour suits you, i.e. #4403?
@AnotherLife @pew-pew could you test if there are no more issues with playback breaking after opening details, i.e. #4541?
@aand18 @Medard22 @UserX404 @ventilaar @floral-qua-floral @Idiotten could you test if this solves #4432, i.e. background playback stopping at random after some (constant) time?
[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5476731/app-debug.zip)

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Medard22 commented:
So far seems it seems good. I didn´t encounter any problems with the playback. I will test it thoroughly tommorow.

ventilaar commented:
Tested for issue #4432, fixed for me. Thank you

avently commented:
> Disable player stream preloading only if the current stream is going to be replaced for sure (see this ) @avently is this ok?

I don't think so. 
Why did you add all those checks?
```java
newQueue != null && playQueue != null
!Objects.equals(newQueue.getItem(), playQueue.getItem())
```

What's the reason for them to exist? They will not fix the issue with preloading of background & popup because the fix would be to apply disabled preloading for `player.videoPlayerSelected()`. In your case you probably will see a situation when `Objects.equals(newQueue.getItem(), playQueue.getItem())` is always true even for the same streams because of different restore position (or maybe I'm wrong, don't remember the contents of playqueueitem). But it will not fix the problem with preloading anyway.

Why did you delete lines after https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java#L858 ? They were here for a reason. If you found another way to do the same thing, then ok.

You made here many good things (based on description). It's hard to see all new code on a phone but fixes and futures you introduced is very important!

avently commented:
Ok. Some things I found:
- the result of removing the code I mentioned and objects.equals checks:
Open any video in audio player
Tap on notification
Choose to play in main player in three dots menu
Player is not changed to main, still audio only (problem)
If you tap on notification and choose another time on timeline - endless loading - another problem.

- try to load some videos one by one (using related videos), play some of them, then try to press back and (with autorotation enabled) rotate the phone into another orientation. Try this step with backpress and orientation change multiple times and you'll see that after rotation there is another video (not the one before rotation). It's probably because if removing static modifiers or maybe it was like this even in 0.20.0, didn't check.

AnotherLife commented:
The "popup player stopping after viewing video details" problem does not happen with this apk from what I tested.

Idiotten commented:
I have not encountered #4432 after a couple hours of queuing and such. Seems fixed for me.

opusforlife2 commented:
Yay! So many good changes! Gonna test now.

opusforlife2 commented:
Does this not include the decryption retry code? I got that error just now.

TobiGr commented:
@opusforlife2 This branch is not based on the latest dev. that fix is not included

opusforlife2 commented:
Can confirm it fixes #4403. 👍

MD77MD commented:
>When sharing a video to NewPipe and a play button is pressed, replace the current queue instead of enqueueing. Enqueueing can still be achieved opening info and then enqueueing, but enqueueing directly was bad UX since the user would not see what happened.

this is great if share to player, but please make it that it sould enqueue when sharing to pup-up or background player.... or a separate enqueue option

Stypox commented:
> This is the only way to enqueue videos from other apps to newPipe

As I explained, this is not true, since now if there is a video already playing the "Show info" button is shown, and from video details you can just long press on background or popup

Stypox commented:
@avently About the preloading part, I then am not sure I understand when preloading should be disabled. From what I understood so far: when the current stream is being dismissed in favour of a new one, then preloading can be disabled for that track since it is soon going to be replaced. Therefore, when switching queues, we should make sure the newly selected video is different from the current one, otherwise preloading would be disabled forever. Ok, maybe if the video is the same and the timestamp differs we could disable preloading, but that's just a special case, so I wouldn't add complexity for that for now. 

> the result of removing the code I mentioned and objects.equals checks:

Do you mean that by removing my conditions for preloading it creates the explained issues, or that my code creates those?

> Why did you delete lines after https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java#L858 ? They were here for a reason. If you found another way to do the same thing, then ok.

If you are talking about "Situation when user switches from players to main player. All needed data is here, we can start watching" then yeah, I am handling the switching players case directly [here](https://github.com/Stypox/NewPipe/blob/fix-detail-open/app/src/main/java/org/schabi/newpipe/util/NavigationHelper.java#L388)

Stypox commented:
Another apk with the same changes but with updated extractor: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5467484/app-debug.zip)

avently commented:
@Stypox here I provided steps for reproduce problems. You said that you made another way for switching players so that way should be used in order to switch to main player from BackgroundActivity like in my reproducible example. Because now you're not switching players when moving to video details from backgroundactivity.

About preloading part:
I think you just need to make sure it's actually works. If it is then current code is ok. But again in my reproducible example the code is not working properly. 

Did you checked with logs, for example, that your code for preloading really getting called and not just skipped when one stream changes to another?

MD77MD commented:
> @MD77MD As I explained, this is not true, since now if there is a video already playing the "Show info" button is shown, and from video details you can just long press on background or popup

@Stypox that is true but please bear with me. lets say you did a Google search in browser and found a number of useful videos... now you have to open the details page of the video then long press to enqueue then go back the search page and do this all over again for each video... can you imagine how much time, effort and data is wasted going back and forth. all off this while the video pup-up is in your face, moving it up and down because its abstracting the button you want to tap.

I'm not criticizing, i really like what your doing. but  i wish if you would reconsider finding a work around for this please . perhaps a tiny window asking to show details or enqueue. 

or we could just use the regular share window but utilizing the new enqueue option from #4425. This would add an extra stab, but a light and easy one.

anyway I'm really thankful to everything you've do here.

opusforlife2 commented:
@MD77MD Set Preferred open action to Always Ask, then long press on background or popup to enqueue.

Stypox commented:
@avently @TobiGr @opusforlife2 @B0pol I added more changes, and fixed the issue reported by avently:
- **also have the popup fullscreen button and the player queue switch-to-main button follow the new method**
- `equals()` was implemented for PlayQueueItems, so that (only) the url is compared when checking them.
- Make sure play/pause state is preserved when switching players via the specific buttons. Also replace `START_PAUSED` with `PLAY_WHEN_READY` for better consistency (i.e. not inverting twice).
- Move menu listener to switch players from `BackgroundPlayerActivity` to `ServicePlayerActivity`, thus removing a useless abstract function.
- Refactor `getPlayerIntent()` functions in `NavigationHelper` and removed unused navigation functions.
- Fix some warnings in `PlayerHelper`

[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5469727/app-debug.zip)

avently commented:
Changes sound good, the code for new commit looks good too. I always like optimizations:)

MD77MD commented:
> @MD77MD Set Preferred open action to Always Ask, then long press on background or popup to enqueue.

@opusforlife2 I did what you said, the menu did show up but when long pressing on the pop-up option nothing happens.... I wish this would work but it's not working... could you try it yourself. i used the apk from stypox last comment.

opusforlife2 commented:
@MD77MD Apparently, you need to tap the option once so that it is highlighted, then long press and leave it. A bit unintuitive, which should be improved.

opusforlife2 commented:
@Stypox 2 of the 3 linked issues are reported to be fixed. What else needs to be tested, exactly?

Stypox commented:
I guess a last round of tests about the things I changed would be enough, just to be sure the new behaviour is not broken. Other than that, yeah, I think this is ready

opusforlife2 commented:
> I guess a last round of tests about the things I changed would be enough

@Stypox But... but the stuff you've changed is all programming jargon. :( That's why I don't know what to test.

Stypox commented:
@opusforlife2 just opening the video detail fragment under many different conditions is enough

opusforlife2 commented:
> opening urls in descriptions in NewPipe now does not erase backstack, nor does clicking Open details on queue items, and probably something else. :-D

Yay!

opusforlife2 commented:
@Stypox With Autoplay off, tapping on the thumbnail to play is broken now. It opens a notification which buffers forever, then does nothing but stay on the screen. If you open a related video and tap the thumbnail, it says "could not play this stream".

If you go back to the previous video and tap the thumbnail, this time it opens the video, but paused at the beginning. Manually tapping the play button finally makes it play.

Stypox commented:
@opusforlife2 fixed, thank you for testing
@leewp14 I added a check that opens the mini player on app open if there is a player running in background/popup. Could you test if this apk solves #4712?
[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5471719/app-debug.zip)

MD77MD commented:
> @MD77MD Apparently, you need to tap the option once so that it is highlighted, then long press and leave it. A bit unintuitive, which should be improved.

just to be clear we are at the same page, long pressing is supposed to enqueue video. if so, in my case long pressing does nothing, and when i release my finger it just replace the already playing video and doesn't enqueue. 

if it not to much to ask, could you make a small video of enqueueing two or three videos in pop-up mode. maybe im doing something wrong. 

Anyway, thank you @opusforlife2 for trying to help. i really appreciate it 🌷😊

opusforlife2 commented:
@Stypox Confirmed fixed.

> I added a check that opens the mini player on app open if there is a player running in background/popup.

That was really bugging me! Thanks! It even works if you have nothing playing but the mini player is still visible, like when you have video details open. You shouldn't have to lose your backstack just by an accidental app switch.

opusforlife2 commented:
@MD77MD If you could come on IRC this would go much, _much_ faster. Make a Matrix account and join #freenode_#newpipe:matrix.org, dude.

Are you trying to long press on Background or Popup? Those are the only two where this works.

Stypox commented:
An apk containing both VideoDetailFragment fixes TeamNewPipe/NewPipe#4562 and deobfuscation improvements TeamNewPipe/NewPipeExtractor#447, in case someone (@opusforlife2) wants to test both at the same time: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5473958/app-debug.zip)

Medard22 commented:
> An apk containing both VideoDetailFragment fixes #4562 and deobfuscation improvements [TeamNewPipe/NewPipeExtractor#447](https://github.com/TeamNewPipe/NewPipeExtractor/pull/447), in case someone (@opusforlife2) wants to test both at the same time: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5473958/app-debug.zip)

This build is consistently crashing on my device when going from popup player to fullscreen. 

![Screenshot_20201102-150441](https://user-images.githubusercontent.com/42897361/97877069-f6182780-1d1c-11eb-903e-f32a783e8a84.png)

Stypox commented:
@Medard22 thank you for testing! Does this apk fix the issue? @opusforlife2 is also does not contain better-deobfuscation anymore.
[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5476731/app-debug.zip)

Medard22 commented:
@Stypox Yes it is fixed now. Thanks.

Also I noticed one thing, but I am not sure if it is directly related to this issue. When I have a queue in the background player and then run any video in the foreground the queue will be emptied. Is that intended? Sorry for the possible offtopic.

opusforlife2 commented:
@Medard22 There is an option in the settings to warn you in case that happens, letting you choose what to do.

Medard22 commented:
@opusforlife2 Oops, I totally missed that. Thank you.

opusforlife2 commented:
Bug: Volume and brightness gesture areas are swapped in this APK. Volume is on the left, but it should be on the right.

Stypox commented:
Why? I didn't touch anything about that... it is probably caused by the recently merged improvements to gestures by @vkay94 and should be a really small fix (i.e. inverting an if)

opusforlife2 commented:
Yeah, it's probably that. I didn't test that PR.

vkay94 commented:
@Stypox The swap lies in my PR indeed :( 
But it can be fixed easily here: https://github.com/Stypox/NewPipe/blob/97f149b0da4c02518496b46fb3f6a79f8ecaaad6/app/src/main/java/org/schabi/newpipe/player/event/PlayerGestureListener.java#L112

Sorry,

Stypox commented:
@vkay94 there is nothing to be sorry about, accidental problems always happen ;-)

vkay94 commented:
Yeah, I know, but I through I checked it carefully :) @Stypox Can you add the fix with this PR? Simply replace both methods with each other ;)

Stypox commented:
@vkay94 is this ok?
@opusforlife2 [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5488555/app-debug.zip)

opusforlife2 commented:
Fixed.

Stypox commented:
I rebased and tested the rebased version all day with multiple situations and everything worked flawlessly: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5506813/app-debug.zip)

The only issue I found, that is present even in NewPipe 0.20.2, and is thus not caused by this PR, is:
- enqueue a random video in the background
- open a local playlist, a remote playlist, a channel or the history fragment
- tap on "Play all"
- the main player does not start playing right away (as it would be expected), instead the video details of the first item of the playlist are shown without playing and the rest of the queue is not preserved

Stypox commented:
I rebased and tested the rebased version all day with multiple situations and everything worked flawlessly: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5506813/app-debug.zip)

The only issue I found, that is present even in NewPipe 0.20.2, and is thus not caused by this PR, is:
- enqueue a random video in the background
- open a local playlist, a remote playlist, a channel or the history fragment
- tap on "Play all"
- the main player does not start playing right away (as it would be expected), instead the video details of the first item of the playlist are shown without playing and the rest of the queue is not preserved

Since the above issue is not caused by this PR I will take care of fixing it in a future one, so I think this can be merged @TobiGr

B0pol commented:
> The only issue I found, that is present even in NewPipe 0.20.2, and is thus not caused by this PR

Then please open a new issue

Stypox commented:
@B0pol I was going to ;-)

jay00128 commented:
#4678. my issue ：video names on Youtube are Chinese，but on NewPipe they became english titles，maybe it's because of the community contributions 。🤔

Stypox commented:
@jay00128 please open an issue in english about what you are experiencing

89z commented:
https://github.com/TeamNewPipe/NewPipe/issues/5530
https://github.com/TeamNewPipe/NewPipe/issues/4779

https://github.com/TeamNewPipe/NewPipe/pull/4562

89z commented:
https://github.com/TeamNewPipe/NewPipe/issues/4779
https://github.com/TeamNewPipe/NewPipe/issues/5541

https://github.com/TeamNewPipe/NewPipe/pull/4562
